sqld/src/database/write_proxy/replication.rs

```diff
@@ -207,43 +200,149 @@ fn free_page_header(h: *const PgHdr) {
+            .as_u128();
+        let generation_id = Uuid::from_str(&hello.generation_id)
+            .context("invalid generation id from primary")?
+            .as_u128();
+
+        Ok(Self {
+            pre_commit_index: 0,
+            post_commit_index: 0,
+            generation_id,
+            database_id,
+        })
+    }
+}
+
 impl ReadReplicationHook {
     async fn new(logger: WalLogClient<Channel>, db_path: &Path) -> anyhow::Result<Self> {
         let path = db_path.join("client_wal_index");
-        let last_applied_index_file = OpenOptions::new()
+        let index_meta_file = OpenOptions::new()
             .create(true)
             .read(true)
             .write(true)
             .open(path)?;
 
-        let mut buf = [0; 2 * size_of::<u64>()];
-        let last_applied_index = match last_applied_index_file.read_exact_at(&mut buf, 0) {
-            Ok(()) => {
-                let pre = u64::from_le_bytes(buf[..size_of::<u64>()].try_into()?);
-                let post = u64::from_le_bytes(buf[size_of::<u64>()..].try_into()?);
-                ensure!(pre == post, "database corrupted");
-                Some(pre)
-            }
-            Err(e) if e.kind() == ErrorKind::UnexpectedEof => {
-                // OK: the file was empty, we don't have a log yet.
-                None
-            }
-            Err(e) => Err(e)?,
-        };
+        let meta = WalIndexMeta::read(&index_meta_file)?;
+        let wal_index_meta = Self::perform_handshake(meta, logger.clone()).await?;
 
         Ok(Self {
             logger,
-            // ask for the frame right after the one we last applied
-            fetch_frame_index: last_applied_index.map(|x| x + 1).unwrap_or_default(),
-            last_applied_index_file,
-            last_applied_index,
+            wal_index_meta_file: index_meta_file,
+            wal_index_meta,
             buffer: Default::default(),
             rt: Handle::current(),
-            // TODO: read from file
-            current_gen: 0,
         })
     }
 
+    async fn perform_handshake(
+        meta: Option<WalIndexMeta>,
+        mut logger: WalLogClient<Channel>,
+    ) -> anyhow::Result<WalIndexMeta> {
+        let hello = logger.hello(HelloRequest {}).await?.into_inner();
+        match meta {
+            Some(meta) => match meta.merge_from_hello(hello) {
+                Ok(meta) => Ok(meta),
+                Err(e @ MetaMergeError::Lagging) => {
+                    tracing::error!("Replica ahead of primary: hard-reseting replica");
+                    HARD_RESET.notify_waiters();
+                    return Err(e.into());
+                }
+                Err(e) => return Err(e.into()),
+            },
+            None => Ok(WalIndexMeta::new_from_hello(hello)?),
+        }
+    }
+
+    fn flush_meta(&self) -> anyhow::Result<()> {
+        self.wal_index_meta_file
+            .write_all_at(bytes_of(&self.wal_index_meta), 0)?;
+
+        Ok(())
+    }
```

Comment:
I'm just wondering, shouldn't this be async? It seems to be disk IO.

Reply:
it is called from a sync context

Reply:
I see.